src/conf.c

```diff
@@ -20,509 +20,551 @@
+		chacha_poly1305_set_key(ctx, mesh->config_key);
+		bool success = false;
+
+		if(chacha_poly1305_encrypt_iv96(ctx, seqbuf, config->buf, config->len, buf, &len)) {
+			success = fwrite(seqbuf, sizeof(seqbuf), 1, f) == 1 && fwrite(buf, len, 1, f) == 1;
+		} else {
+			logger(mesh, MESHLINK_ERROR, "Cannot encrypt config file\n");
+			meshlink_errno = MESHLINK_ESTORAGE;
+		}
 
-	if(ai) {
-		*result = ai;
-		return true;
+		chacha_poly1305_exit(ctx);
+		return success;
 	}
 
-	logger(NULL, MESHLINK_ERROR, "Hostname or IP address expected for configuration variable %s in line %d",
-	       cfg->variable, cfg->line);
+	if(fwrite(config->buf, config->len, 1, f) != 1) {
+		logger(mesh, MESHLINK_ERROR, "Cannot write config file: %s", strerror(errno));
+		meshlink_errno = MESHLINK_ESTORAGE;
+		return false;
+	}
 
-	return false;
+	return true;
 }
 
-/*
-  Read exactly one line and strip the trailing newline if any.
-*/
-static char *readline(FILE *fp, char *buf, size_t buflen) {
-	char *newline = NULL;
-	char *p;
+/// Free resources of a loaded configuration file.
+void config_free(config_t *config) {
+	free((uint8_t *)config->buf);
+	config->buf = NULL;
+	config->len = 0;
+}
 
-	if(feof(fp)) {
-		return NULL;
+/// Check the presence of a host configuration file.
+bool config_exists(meshlink_handle_t *mesh, const char *name) {
+	if(!mesh->confbase) {
+		return false;
 	}
 
-	p = fgets(buf, buflen, fp);
+	char path[PATH_MAX];
+	make_host_path(mesh, name, path, sizeof(path));
 
-	if(!p) {
-		return NULL;
+	return access(path, F_OK) == 0;
+}
+
+/// Read a host configuration file.
+bool config_read(meshlink_handle_t *mesh, const char *name, config_t *config) {
+	if(!mesh->confbase) {
+		return false;
 	}
 
-	newline = strchr(p, '\n');
+	char path[PATH_MAX];
+	make_host_path(mesh, name, path, sizeof(path));
 
-	if(!newline) {
-		return buf;
-	}
+	FILE *f = fopen(path, "r");
 
-	/* kill newline and carriage return if necessary */
-	*newline = '\0';
+	if(!f) {
+		logger(mesh, MESHLINK_ERROR, "Failed to open `%s': %s", path, strerror(errno));
+		return false;
+	}
 
-	if(newline > p && newline[-1] == '\r') {
-		newline[-1] = '\0';
+	if(!config_read_file(mesh, f, config)) {
+		logger(mesh, MESHLINK_ERROR, "Failed to read `%s': %s", path, strerror(errno));
+		fclose(f);
+		return false;
 	}
 
-	return buf;
+	fclose(f);
+	return true;
 }
 
-config_t *parse_config_line(char *line, const char *fname, int lineno) {
-	config_t *cfg;
-	int len;
-	char *variable, *value, *eol;
-	variable = value = line;
-
-	eol = line + strlen(line);
-
-	while(strchr("\t ", *--eol)) {
-		*eol = '\0';
+void config_scan_all(meshlink_handle_t *mesh, config_scan_action_t action) {
+	if(!mesh->confbase) {
+		return;
 	}
 
-	len = strcspn(value, "\t =");
-	value += len;
-	value += strspn(value, "\t ");
-
-	if(*value == '=') {
-		value++;
-		value += strspn(value, "\t ");
-	}
+	DIR *dir;
+	struct dirent *ent;
+	char dname[PATH_MAX];
+	make_host_path(mesh, NULL, dname, sizeof(dname));
```

Comment:
We are passing `NULL` as host file name rather than `""` or null character to `make_host_path()` which makes the mesh not to read any of the node files here.

Reply:
Guus you missed this fix.